Parachains db column "migration"

sandreim opened the pull request:
This is a followup of https://github.com/paritytech/polkadot/pull/5594 . This aims to be a "catch" all migration for any column option changes such that no manual fixing for ParityDB is required. Should also work if client is rolled back.

sandreim commented:
Tested this locally, but a Versi burn in is needed to see how much time migration takes and if there are any issues afterwards

sandreim commented:
> The only issue was that paritydb has stored unordered columns as hashes only with no values, so @eskimor told that there is no way to migrate with data preservation. Have you tried that such a migration does really work?

Not sure what happens to the data right now, it shouldn’t be preserved. Its a soft wipeout and hopefully there is no dangling stuff. The good partt is that there is no startup error now and no need to manually ‘fix’ the db.

arkpar commented:
I don't think `migrate` works correctly for moving data from hash column to btree.
Unless the previous version set the `uniform` flag for the colum, there's no way to recover the original key, as the database does not store it.
@cheme  Please correct me if I'm wrong here.

Why not just delete the parachains database instead? As far as I know it should be re-populated in a matter of a few blocks.

cheme commented:
Can only confirm, only 'uniform' column can be migrated.

cheme commented:
Migrate function here should work when changing configuration of hash indexed column (hash indexed to hash indexed), and also with the same parameters (can gain reclaim some space sometime).

When migrating from hash indexed to btree, what may currently happen is that it migrate the index and values tables (the btree_indexed option parameter is currently ignored by the migration). After that, probably opening the db with btree_index act as if the column was deleted (no entry may have been written where the btree root usually is) and things may run fine.

For btree indexed column the migrate function is `unimplemented`, but should be easy to implement (just run the iterator on the whole db and inject all key values to the target column) and would allow to switch from btree indexed to hash indexed.

The fact that we cannot recover key of the content stored in column may be a concern in some case, idea should be to attach key to value then or implement a variant that keep original key along the value (or in an attached value entry). But that is a different question.

Also depending on context if a single column needs to be removed, deleting the metadata file and all index_X_* and table_X_* with X the column number should work (almost untested).

vstakhov commented:
> Why not just delete the parachains database instead?

It would be enough to delete dispute coordinator column I suppose. The goal here is to provide more or less smooth migration. Even a shell script that performs `rm -fr` is better than crashing on an incorrect database.

sandreim commented:
I was hoping for a more sane solution especially on the user experience. Automated db removal sounds a bit scary to me.

cheme commented:
+1 on the fact that automating a column removal is scary, maybe if we know some column are allowed to be removed we can hardcode the info and only run this on such column.
But more generally, I think a warning should be issued before (maybe even a confirmation): synching the data can take time and users should know the node may be a bit non responsive.

sandreim commented:
> +1 on the fact that automating a column removal is scary, maybe if we know some column are allowed to be removed we can hardcode the info and only run this on such column. But more generally, I think a warning should be issued before (maybe even a confirmation): synching the data can take time and users should know the node may be a bit non responsive.

I've implemented it to work only for 1 column (dispute coordinator data) and will spit out warnings. There shouldn't be any issues with sync and responsiveness AFAIK.

arkpar commented:
Database implementation details w.r.t. file names may change in the future and this will break.  There should be a function for clearing a column provided by the database.  I've added one here:
https://github.com/paritytech/parity-db/pull/84

sandreim commented:
Thanks, you are right, this relies on impl details. I’d expand on that - move most of this code in paritydb and add an additional column option to delete it when no migration is available. Would this make work?

arkpar commented:
Not sure this is a good idea. Mixing actual configuration options with some flags that specify behaviour for some edge cases. I'd rather have it simple: try to open the database in the user code. If there's a particular error, the user code may handle it by dropping the column. I've added an explicit error type `IncompatibleColumnConfig` to make it possible to detect it better.

ordian commented on node/service/Cargo.toml:
```diff
@@ -64,12 +64,13 @@ prometheus-endpoint = { package = "substrate-prometheus-endpoint", git = "https:
 futures = "0.3.21"
 hex-literal = "0.3.4"
 gum = { package = "tracing-gum", path = "../gum/" }
+glob = "0.3.0"
```
unused?
```suggestion
```

ordian reviewed (commented):
IMHO we should try to clear a column only if opening the db fails with some specific error. Other than that, lgtm.

sandreim commented:
> IMHO we should try to clear a column only if opening the db fails with some specific error. Other than that, lgtm.

I think this is currently more efficient and gives more control on how the issue is resolved, because we can detect the condition ourselves, and doesn't require to have a loop in which to open and delete on error. If you have a strong opinion about this I would change it.

ordian commented on node/service/src/parachains_db/mod.rs:
```diff
@@ -138,6 +196,8 @@ pub fn open_creating_paritydb(
 		options.columns[*i as usize].btree_index = true;
 	}
 
+	fix_columns(&path, options.clone(), vec![columns::COL_DISPUTE_COORDINATOR_DATA])?;
```
We already have versioning in https://github.com/paritytech/polkadot/blob/dd256a327c2d6e7d9f941362ada0cce583d1d003/node/service/src/parachains_db/upgrade.rs#L30 unless you mean something else. The problem is that it was assumed that changing a setting of a column is a backwards compatible change no version was bumped.

bkchr replied:
Yeah I meant such a version, but it was already known for the release that we needed to delete the db. There we should have bumped the version and handled that accordingly.

sandreim replied:
Ok, this makes total sense to have the version bumped, but we need to also be able to roll back as a safety mechanism in the event of some other bug. This is the reason for the more generic approach - should work across versions since we actually check the diffs between db col options. @ordian point is fair wrt using parity db to detect that condition as well. I’ll try to see how we can make this work with the version number here as well. Ideally users should not be bothered with making manual changes to db storage.

sandreim replied:
> We already have versioning in
> 
> https://github.com/paritytech/polkadot/blob/dd256a327c2d6e7d9f941362ada0cce583d1d003/node/service/src/parachains_db/upgrade.rs#L30
> 
> unless you mean something else. The problem is that it was assumed that changing a setting of a column is a backwards compatible change no version was bumped.

This doesn't seem to be implemented for ParityDB.

ordian commented:
> > IMHO we should try to clear a column only if opening the db fails with some specific error. Other than that, lgtm.
> 
> I think this is currently more efficient and gives more control on how the issue is resolved, because we can detect the condition ourselves, and doesn't require to have a loop in which to open and delete on error. If you have a strong opinion about this I would change it.

It doesn't need to be a loop, just opening at most 2 times (regular open, and in case of this specific error, try clearing a column and open again (w/o a retry), otherwise, propagate the error). My main worry is that generally, we don't want to wipe the parachains db (or part of it) on all nodes, esp once paritydb replaces rocksdb and we do this on more than 1/3 of validators. This could lead to security and/or consensus issues.

sandreim commented:
> > > IMHO we should try to clear a column only if opening the db fails with some specific error. Other than that, lgtm.
> > 
> > 
> > I think this is currently more efficient and gives more control on how the issue is resolved, because we can detect the condition ourselves, and doesn't require to have a loop in which to open and delete on error. If you have a strong opinion about this I would change it.
> 
> It doesn't need to be a loop, just opening at most 2 times (regular open, and in case of this specific error, try clearing a column and open again (w/o a retry), otherwise, propagate the error). My main worry is that generally, we don't want to wipe the parachains db (or part of it) on all nodes, esp once paritydb replaces rocksdb and we do this on more than 1/3 of validators. This could lead to security and/or consensus issues.

Assuming one column needs clearing, yeah 2 db opens would work, but if we don;t really know the number of cols which need fixing we need to loop and err on each col index and then fix it. If we use the version file bumper, then we would need manually hardcode the migration explicitly like it works for rocks db.

sandreim commented:
I've implemented the versioning for ParityDB which was missing and added the column clear code as a parityDB migration from anything in the past to current version.

sandreim commented:
@bkchr @ordian PTAL

ordian commented on node/service/src/parachains_db/upgrade.rs:
```diff
@@ -102,3 +125,79 @@ fn migrate_from_version_0_to_1(path: &Path) -> Result<(), Error> {
+	let mut options = parity_db::Options::with_columns(&path, super::columns::NUM_COLUMNS as u8);
+	for i in columns::ORDERED_COL {
+		options.columns[*i as usize].btree_index = true;
+	}
```
can we extract the option creation into a function to dedup with `open_creating_paritydb` in case it changes later?

ordian commented on node/service/src/parachains_db/upgrade.rs:
```diff
@@ -102,3 +125,79 @@ fn migrate_from_version_0_to_1(path: &Path) -> Result<(), Error> {
+	// Delete the `dispute coordinator` column if needed (if column configuration is changed).
+	paritydb_fix_columns(
+		&path,
+		options.clone(),
```
```suggestion
		options,
```

ordian commented on node/service/src/parachains_db/upgrade.rs:
```diff
@@ -49,26 +50,39 @@ impl From<Error> for io::Error {
+			// We handle this as a `0 -> 1` migration.
+			None if db_kind == DatabaseKind::ParityDB =>
+				migrate_from_version_0_to_1(db_path, db_kind)?,
+			// Fresh db.
```
what do you mean by "Fresh db"? :) this is unreachable

sandreim replied:
lol

ordian commented on node/service/src/parachains_db/upgrade.rs:
```diff
@@ -102,3 +125,79 @@ fn migrate_from_version_0_to_1(path: &Path) -> Result<(), Error> {
+) -> io::Result<()> {
+	// Figure out which columns to delete. This will be determined by inspecting
+	// the metadata file.
+	if let Ok(Some(metadata)) = parity_db::Options::load_metadata(&path) {
```
should we propagate the error if we loading metadata errors?

sandreim replied:
I think you are right, now this would only be called if the db exists.

ordian commented on node/service/src/parachains_db/upgrade.rs:
```diff
@@ -86,9 +100,18 @@ fn version_file_path(path: &Path) -> PathBuf {
 }
 
+fn migrate_from_version_0_to_1(path: &Path, db_kind: DatabaseKind) -> Result<(), Error> {
+	gum::info!(target: LOG_TARGET, "Migrating parachains db from version 0 to version 1",);
```
would be nice to print "Migration completed successfully" on completion

ordian reviewed (approved):
We should test this, but looks reasonable.

bkchr commented on node/service/src/parachains_db/upgrade.rs:
```diff
@@ -49,26 +50,38 @@ impl From<Error> for io::Error {
+			Some(CURRENT_VERSION) => (),
+			// This is an arbitrary future version, we don't handle it.
+			Some(v) => return Err(Error::FutureVersion { current: CURRENT_VERSION, got: v }),
+			// No version file. For `RocksDB` we dont need to do anything.
+			None if db_kind == DatabaseKind::RocksDB => (),
```
Why don't we need to do anything for rocksdb?

ordian replied:
If no db version found, it probably means there's no db, so we should just write the current version, which we do below.

bkchr replied:
As it was there before, shouldn't it had also already being created for parity db?

sandreim replied:
The version file was not created previously for ParityDB, seems that this was somehow missed.

ordian replied:
Previously, we only called `upgrade::try_upgrade_db` for RocksDB, not ParityDB. So ParityDB's parachain db didn't have any versioning. That was an oversight.

ordian commented on node/service/src/parachains_db/upgrade.rs:
```diff
@@ -102,3 +128,131 @@ fn migrate_from_version_0_to_1(path: &Path) -> Result<(), Error> {
+			.unwrap();
+		}
+
+		try_upgrade_db(&path, DatabaseKind::ParityDB).unwrap();
+
+		let db = Db::open(&paritydb_version_1_config(&path)).unwrap();
```
Thanks for adding a test. Can we use `open_creating_paritydb` instead?

ordian replied:
Actually, might be better idea to leave as is. This tests opening with v1, not the current version.

sandreim commented:
bot merge
